core/field_textdropdown.js

```diff
@@ -0,0 +1,149 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/**
+ * @fileoverview Combination text + drop-down field
+ * @author tmickel@mit.edu (Tim Mickel)
+ */
+'use strict';
+
+goog.provide('Blockly.FieldTextDropdown');
+
+goog.require('Blockly.DropDownDiv');
+goog.require('Blockly.FieldDropdown');
+goog.require('Blockly.FieldTextInput');
+goog.require('goog.userAgent');
+
+
+/**
+ * Class for a combination text + drop-down field.
+ * @param {string} text The initial content of the text field.
+ * @param {(!Array.<!Array.<string>>|!Function)} menuGenerator An array of
+ *     options for a dropdown list, or a function which generates these options.
+ * @param {Function=} opt_validator An optional function that is called
+ *     to validate any constraints on what the user entered.  Takes the new
+ *     text as an argument and returns the accepted text or null to abort
+ *     the change.
+ * @param {RegExp=} opt_restrictor An optional regular expression to restrict
+ *     typed text to. Text that doesn't match the restrictor will never show
+ *     in the text field.
+ * @extends {Blockly.FieldTextInput}
+ * @constructor
+ */
+Blockly.FieldTextDropdown = function(text, menuGenerator, opt_validator, opt_restrictor) {
+  this.menuGenerator_ = menuGenerator;
+  Blockly.FieldDropdown.prototype.trimOptions_.call(this);
+  Blockly.FieldTextDropdown.superClass_.constructor.call(this, text, opt_validator, opt_restrictor);
+  this.addArgType('textdropdown');
+};
+goog.inherits(Blockly.FieldTextDropdown, Blockly.FieldTextInput);
+
+/**
+ * Install this text drop-down field on a block.
+ */
+Blockly.FieldTextDropdown.prototype.init = function() {
+  Blockly.FieldTextDropdown.superClass_.init.call(this);
+  // Add dropdown arrow: "option ▾" (LTR) or "▾ אופציה" (RTL)
+  // Positioned on render, after text size is calculated.
+  if (!this.arrow_) {
+    /** @type {Number} */
+    this.arrowSize_ = 12;
+    /** @type {Number} */
+    this.arrowX_ = 0;
+    /** @type {Number} */
+    this.arrowY_ = 11;
+    this.arrow_ = Blockly.createSvgElement('image', {
+      'height': this.arrowSize_ + 'px',
+      'width': this.arrowSize_ + 'px'
+    });
+    this.arrow_.setAttributeNS('http://www.w3.org/1999/xlink',
+        'xlink:href', Blockly.mainWorkspace.options.pathToMedia + 'dropdown-arrow-dark.svg');
+    this.arrow_.style.cursor = 'pointer';
+    this.fieldGroup_.appendChild(this.arrow_);
+    this.mouseUpWrapper_ =
+        Blockly.bindEvent_(this.arrow_, 'mouseup', this,
+            this.showDropdown_);
+  }
+  // Prevent the drop-down handler from changing the field colour on open.
+  this.disableColourChange_ = true;
+};
+
+/**
+ * Close the input widget if this input is being deleted.
+ */
+Blockly.FieldTextDropdown.prototype.dispose = function() {
+  if (this.mouseUpWrapper_) {
+    Blockly.unbindEvent_(this.mouseUpWrapper_);
```

Comment:
Also need to call Blockly.Touch.clearTouchIdentifier() when ending a touch, or you'll get stuck (most obvious on ios).